### ISSUE TITLE
Traefik is serving default TLS certificate during ACME/TLS-ALPN-01 challenge

### ISSUE THREAD

**Issue by Kanshiroron** [containous/traefik#4850]
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [x] No

### What did you do?

Traefik version (docker tag): traefik:1.7-alpine

I have configured Traefik on Kubernetes with following configuration (K8s/Etcd/ACME):

```toml
    # ENTRYPOINTS
    defaultEntryPoints = ["https", "http"] # !!! This has to be at the top of the file
    [entryPoints]
      [entryPoints.http]
        address = ":80"
        [entryPoints.http.redirect]
          entryPoint = "https"
      [entryPoints.https]
        address = ":443"
        [entryPoints.https.tls]
          minVersion = "VersionTLS12"
          cipherSuites = [
            "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
            "TLS_RSA_WITH_AES_256_GCM_SHA384"
          ]
      [entryPoints.monitoring]
        address = ":8080"

    # LOGS
    logLevel = "WARN"

    [traefikLog]
      format   = "json"

    # ACME
    [acme]
    email = "me@example.com"
    storage = "traefik/acme/account"
    entryPoint = "https"
    onHostRule = true
    caServer = "https://acme-staging-v02.api.letsencrypt.org/directory" # for testing
    [acme.tlsChallenge]
    
    # KUBERNETES
    [kubernetes]
    # Value of `kubernetes.io/ingress.class` annotation that identifies Ingress objects to be processed.
    # If the parameter is non-empty, only Ingresses containing an annotation with the same value are processed.
    # Otherwise, Ingresses missing the annotation, having an empty value, or the value `traefik` are processed.
    ingressClass = "traefik"

    [etcd] # to store Let's Encrypt certificates
    endpoint = "etcd:2379"
    watch = true
    prefix = "/traefik"
    useAPIV3 = true

    [respondingTimeouts]
    # readTimeout is the maximum duration for reading the entire request, including the body.
    # readTimeout = "5s" # Default: "0s"

    # writeTimeout is the maximum duration before timing out writes of the response.
    writeTimeout = "30s" # Default: "0s"

    # idleTimeout is the maximum duration an idle (keep-alive) connection will remain idle before closing itself.
    idleTimeout = "300s" # Default: "180s"

    # [forwardingTimeouts]
    # dialTimeout is the amount of time to wait until a connection to a backend server can be established.
    # dialTimeout = "30s" # Default: "30s"

    # responseHeaderTimeout is the amount of time to wait for a server's response headers after fully writing the request (including its body, if any).
    # responseHeaderTimeout = "0s" # Default: "0s"

    # ADMIN/MONITORING
    [api]
      # Web administration port.
      entryPoint = "monitoring"
      dashboard = true

    [metrics]
      [metrics.prometheus]
        entryPoint = "monitoring"
        # buckets = [0.1,0.3,1.2,5.0]
```
But when Traefik is trying to get a Let's Encrypt certificate it serves back the default certificate.
```
{"level":"error","msg":"Error getting ACME certificates [test.example.com] : cannot obtain certificates: acme: Error -\u003e One or more domains had a problem:\n[test.example.com] acme: error: 403 :: urn:ietf:params:acme:error:unauthorized :: Incorrect validation certificate for tls-alpn-01 challenge. Requested test.example.com from W.X.Y.Z:443. Received 1 certificate(s), first certificate had names \"32bf782dfd46e7cc3c7eb9618644a519.ddd08ec150005d4b70a05f6fc09250f4.traefik.default, traefik default cert\", url: \n","time":"2019-05-07T03:13:17Z"}
```

However, if I change the K8s ingress definition for my deployment to using a certificate as a secret this is working fine.

Thanks for your support!

**Comment by Kanshiroron** [comment]
Just a quick update: if removed the etcd configuration and it works correctly. So this is somehow linked to the etcd storage backend.

Basically on the configuration, I removed the `[etcd]` section and changed back the `acme.storage` key to a regular file path.

**Comment by nbaum** [comment]
I am encountering the same situation when using Consul as the KV backend.

It also works for me if I change back to storing the ACME data in a file.

**Comment by Kanshiroron** [comment]
Hey Traefik team,

Was nice to meet you guys at the CNCF.
As discussed, I tried to reproduced this behavior with a simple docker-compose, but I couldn't as I'm facing an issue while Traefik is trying to build the ACME client:

```
traefik_1         | {"level":"error","msg":"Cannot unmarshall private key []","time":"2019-05-27T14:32:33Z"}
traefik_1         | {"level":"error","msg":"Error building ACME client \u0026{Email: Registration:\u003cnil\u003e PrivateKey:[] KeyType: DomainsCertificate:{Certs:[] lock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0}} ChallengeCerts:map[] HTTPChallenge:map[]}: private key was nil","time":"2019-05-27T14:32:33Z"}
```

You can find attached the `docker-compose.yml` and the `traefik.toml`
[compose.zip](https://github.com/containous/traefik/files/3224174/compose.zip)

To reproduce:
- `docker-compose up -d etcd`
- `docker-compose up traefik-config` (and wait for Traefik to exit)
- `docker-compose up -d traefik nginx`
- `docker-compose logs -f traefik`

Help this will help you investigate.

@nbaum I don't know what is your use-case using Consul as KV, on my side I was using it to have a Traefik cluster that share Let's Encrypt certificates. According the Traefik team, this feature will be only available on the Enterprise Edition starting version 2.

**Comment by caarlos0** [comment]
Just happened here with stable/traefik helm chart

`traefik.toml` looks like this:

```toml
# traefik.toml
logLevel = "info"
defaultEntryPoints = ["http","https"]
[entryPoints]
  [entryPoints.http]
  address = ":80"
  compress = true
    [entryPoints.http.redirect]
      regex = "^http://(.*)"
      replacement = "https://$1"
  [entryPoints.https]
  address = ":443"
  compress = true
    [entryPoints.https.tls]
      [[entryPoints.https.tls.certificates]]
      CertFile = "/ssl/tls.crt"
      KeyFile = "/ssl/tls.key"
[kubernetes]
[traefikLog]
  format = "json"
[acme]
email = "noone@mydomain.com"
storage = "/acme/acme.json"
entryPoint = "https"
onHostRule = true
caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
acmeLogging = true
  [acme.tlsChallenge]
[[acme.domains]]
    main = "foo.mydomain.com"
```

when going to `foo.mydomain.com`, it is serving a cert for `*.example.com`... and of course the validation fails...

We have a single ingress using that same `foo.mydomain.com` as host and it the DNS its a CNAME to an ELB.

Weirdest thing is that it was working at some point, was re-created to test config "from scratch" and never worked anymore...

**Comment by caarlos0** [comment]
PS: after some time (days) left alone it seemed to work.
Same issue happening on a different cluster and domain now though.

also worth mentioning: single replica, using file as backend, deployed with helm, latest versions of both the chart and traefik.

**Comment by svenstucki** [comment]
I've been having the same issue as well. I'm using the helm chart (stable/traefik) with a pretty basic config (acme.enabled, acme.email, acme.challengeType and acme.persistence.enabled) on an AWS EKS cluster.

After seeing this issue, I've tried downgrading the helm chart versions. I get this issue on 1.68.5, 1.68.4 and 1.68.0 (which all correspond to traefik 1.7.9). Then I tried 1.61.0 (highest version with traefik 1.7.7) and it worked!

**Comment by iamliamc** [comment]
I also am having this issue getting things like:
`{"level":"error","msg":"Unable to obtain ACME certificate for domains \"httpd.test.mything.io\" detected thanks to rule \"Host:httpd.test.mything.io\" : unable to generate a certificate for the domains [httpd.test.mything.io]: acme: Error -\u003e One or more domains had a problem:\n[httpd.test.mything.io] acme: error: 403 :: urn:ietf:params:acme:error:unauthorized :: Cannot negotiate ALPN protocol \"acme-tls/1\" for tls-alpn-01 challenge, url: \n","time":"2019-07-01T13:59:02Z"}`

Also a very simple stable/traefik configuration... acme.enabled, acme.email, acme.challengeType, acme.persistence.enabled on a digital oceans cluster... we are trying a downgrade to Traefik 1.7.7 now...

**Comment by caarlos0** [comment]
> I don't know if this means we are by default usint Etcd as storage but I don't think so... so perhaps this issue isn't specific to Etcd...

unless you manually set to use etcd, you're using the default filestore...

**Comment by iamliamc** [comment]
Well all I can say is that when we installed version 1.61.0 of the chart using staging.true we get the appropriate staging cert (“Fake LE Intermediate X1”) in the browser... so i don't think this issue is specific to Etcd...

**Comment by caarlos0** [comment]
I got it with filestore and consul as well...

**Comment by cromega** [comment]
I also get 

```json
{"level":"error","msg":"Unable to obtain ACME certificate for domains \"test.x.y.z\"
 detected thanks to rule \"Host:test.x.y.z\" : unable to generate a certificate for the domains [test.x.y.z]:
 acme: Error -\u003e One or more domains had a problem:\n[test.x.y.z] acme: error: 403 :: 
urn:ietf:params:acme:error:unauthorized :: Incorrect validation certificate for tls-alpn-01 challenge. 
Requested test.x.y.z from 1.2.3.4:443. Received 1 certificate(s), first certificate had names 
\"addcc2cfd65bbf2c3458f6068208dfac.3c7970257c17c4e19ae27881dd9abc6d.traefik.default, traefik 
default cert\", url: \n","time":"2019-07-02T22:46:50Z"}
```

with `storage = "acme.json"`

**Comment by jhburns** [comment]
Can also confirm I'm getting the same error, using Traefik v1.7.12.
Also using etcd backend, on kuberentes.

The ACME config looks like:
```
        - "--acme"
        - "--acme.tlschallenge"
        - "--acme.acmelogging"
        - "--acme.onhostrule"
        - "--acme.email=xxxxxxxx@gmail.com"
        - "--acme.entrypoint=https"
        - "--acme.storage=traefik/acme/account"
        # Please note that this is the staging Let's Encrypt server.
        # Once you get things working, you should remove that whole line altogether.
        - "--acme.caserver=https://acme-staging-v02.api.letsencrypt.org/directory"
```

**Comment by dtomcej** [comment]
The recommendation to downgrade to an older chart version is extremely unwise.

The helm chart is updated with bugfixes for configuration only. It does not have any changes with _how_ acme challenges or KV store interaction is handled. That is handled entirely by the binary within the Traefik container.

Downgrading to an older version of Traefik is also unwise. Traefik version 1.7.12 contains an important bugfix, and is recommended for _ALL_ users. Knowingly using an older version is not great.

You can review the release notes for the versions of traefik (https://github.com/containous/traefik/releases), and you will see that each patch release improves KV interaction, logging, and uses newer versions of the ACME client.

*TL;DR* Use the latest version of the helm chart, and use the latest version of Traefik.

I am going to mark a bunch of comments as off-topic due to them not being the _same_ issue as the OP.

If you want to comment further on this issue, we will need the following:

- An error message that the *WRONG* certificate was presented during the acme challenge (If you don't have this message, this issue isn't your issue)
- A gist containing debug logs. One line of logs is not enough. There are lots of information contained in the logs _around_ the final error message
- What kind of challenge you are using (It should be TLS-ALPN-01, or this isn't your issue)
- How many replicas of Traefik are you using
- What version of ETCD are you using?
- What version of Traefik are you using (The traefik version, not the helm chart version, that doesn't affect the issue)?

Keeping these issues separate, and providing good information is crucial to being able to debug and track down the actual issue.

Thanks everyone!

**Comment by cromega** [comment]
More details following my comment:

Traefik: 1.7.12
ETCD: none
cert storage: file
replicas: 1

logs: https://gist.github.com/cromega/3a5a51dcc629bffbb44bf65aec86bb24

**Comment by asyd** [comment]
Same issue here with traefik v1.7.16. Any update by chance?

**Comment by Maurotb** [comment]
Same issue here with traefik v1.7.12 (docker image) , i'm using regular storage acme.json

**Comment by Maurotb** [comment]
I have try traefik 2.0.1 (docker image) with  acme.json , same problem here...

**Comment by e3b0c442** [comment]
I think I'm running into this issue with k3s v1.17.4+k3s1 which bundles traefik 1.7.19.

Here is the helm chart with the configuration changes I've made:
```
apiVersion: helm.cattle.io/v1
kind: HelmChart
metadata:
  name: traefik
  namespace: kube-system
spec:
  chart: https://%{KUBERNETES_API}%/static/charts/traefik-1.81.0.tgz
  set:
    rbac.enabled: "true"
    ssl.enabled: "true"
    metrics.prometheus.enabled: "true"
    kubernetes.ingressEndpoint.useDefaultPublishedService: "true"
    image: "rancher/library-traefik"
  valuesContent: |-
    acme:
      enabled: true
      email: REDACTED_EMAIL
      onHostRule: false
      logging: true
      challengeType: tls-alpn-01
      domains:
        enabled: true
        domainsList:
          - main: "REDACTED_DOMAIN"
```
Here is the log output from traefik:
```
{"level":"info","msg":"Using TOML configuration file /config/traefik.toml","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"No tls.defaultCertificate given for https: using the first item in tls.certificates as a fallback.","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Traefik version v1.7.19 built on 2019-10-28_02:07:32PM","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"\nStats collection is disabled.\nHelp us improve Traefik by turning this feature on :)\nMore details on: https://docs.traefik.io/v1.7/basics/#collected-data\n","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Preparing server http \u0026{Address::80 TLS:\u003cnil\u003e Redirect:\u003cnil\u003e Auth:\u003cnil\u003e WhitelistSourceRange:[] WhiteList:\u003cnil\u003e Compress:true ProxyProtocol:\u003cnil\u003e ForwardedHeaders:0xc0005bb940} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Preparing server https \u0026{Address::443 TLS:0xc000143320 Redirect:\u003cnil\u003e Auth:\u003cnil\u003e WhitelistSourceRange:[] WhiteList:\u003cnil\u003e Compress:true ProxyProtocol:\u003cnil\u003e ForwardedHeaders:0xc0005bb960} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Preparing server prometheus \u0026{Address::9100 TLS:\u003cnil\u003e Redirect:\u003cnil\u003e Auth:\u003cnil\u003e WhitelistSourceRange:[] WhiteList:\u003cnil\u003e Compress:false ProxyProtocol:\u003cnil\u003e ForwardedHeaders:0xc0005bb9e0} with readTimeout=0s writeTimeout=0s idleTimeout=3m0s","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Starting provider configuration.ProviderAggregator {}","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Starting server on :80","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Starting server on :443","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Starting server on :9100","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Starting provider *kubernetes.Provider {\"Watch\":true,\"Filename\":\"\",\"Constraints\":[],\"Trace\":false,\"TemplateVersion\":0,\"DebugLogGeneratedTemplate\":false,\"Endpoint\":\"\",\"Token\":\"\",\"CertAuthFilePath\":\"\",\"DisablePassHostHeaders\":false,\"EnablePassTLSCert\":false,\"Namespaces\":null,\"LabelSelector\":\"\",\"IngressClass\":\"\",\"IngressEndpoint\":{\"IP\":\"\",\"Hostname\":\"\",\"PublishedService\":\"kube-system/traefik\"},\"ThrottleDuration\":0}","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Starting provider *acme.Provider {\"Email\":\"REDACTED_EMAIL\",\"ACMELogging\":true,\"CAServer\":\"https://acme-staging-v02.api.letsencrypt.org/directory\",\"Storage\":\"/acme/acme.json\",\"EntryPoint\":\"https\",\"KeyType\":\"RSA4096\",\"OnHostRule\":false,\"OnDemand\":false,\"DNSChallenge\":null,\"HTTPChallenge\":null,\"TLSChallenge\":{},\"Domains\":[{\"Main\":\"REDACTED_DOMAIN\",\"SANs\":null}],\"Store\":{}}","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Testing certificate renew...","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"ingress label selector is: \"\"","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Creating in-cluster Provider client","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Server configuration reloaded on :443","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Server configuration reloaded on :9100","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Server configuration reloaded on :80","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"legolog: [INFO] [REDACTED_DOMAIN] acme: Obtaining bundled SAN certificate","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Server configuration reloaded on :80","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Server configuration reloaded on :443","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"Server configuration reloaded on :9100","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"legolog: [INFO] [REDACTED_DOMAIN] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/55825322","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"legolog: [INFO] [REDACTED_DOMAIN] acme: use tls-alpn-01 solver","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"legolog: [INFO] [REDACTED_DOMAIN] acme: Trying to solve TLS-ALPN-01","time":"2020-05-11T13:33:04Z"}
{"level":"info","msg":"legolog: [INFO] Unable to deactivate the authorization: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/55825322","time":"2020-05-11T13:33:09Z"}
{"level":"error","msg":"Unable to obtain ACME certificate for domains \"REDACTED_DOMAIN\" : unable to generate a certificate for the domains [REDACTED_DOMAIN]: acme: Error -\u003e One or more domains had a problem:\n[REDACTED_DOMAIN] acme: error: 403 :: urn:ietf:params:acme:error:unauthorized :: Incorrect validation certificate for tls-alpn-01 challenge. Requested REDACTED_DOMAIN from REDACTED_IP:443. Received 1 certificate(s), first certificate had names \"*.example.com\", url: \n","time":"2020-05-11T13:33:09Z"}

There is an Ingress host for REDACTED_DOMAIN.

Appreciate any insight.

**Comment by spencer-p** [comment]
I experienced the same issue on 1.7.19 as above with k3s version v1.18.9+k3s1 (630bebf9).

I found that deleting the saved certificate solved the problem:

```sh
sudo find -name acme.json -delete
# now restart traefik somehow
```

**Comment by GCSBOSS** [comment]
Having the same issue on 2.2

**Comment by chris-aeviator** [comment]
Having the same issue with the default traefik configuration as soon as moving from http challenge to tlsChallenge

**Comment by mmatur** [comment]
Hello,

This bug is linked to a feature that no longer exists in the supported versions of Traefik.

Closing accordingly.

Please feel free to re-open it if necessary.
